core/drivers/crypto/versal/hash.c

```diff
@@ -0,0 +1,246 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) Foundries Ltd. 2022
+ * Author: Jorge Ramirez <jorge@foundries.io>
+ */
+
+#include <assert.h>
+#include <drvcrypt.h>
+#include <drvcrypt_hash.h>
+#include <initcall.h>
+#include <ipi.h>
+#include <kernel/panic.h>
+#include <kernel/refcount.h>
+#include <kernel/spinlock.h>
+#include <mm/core_memprot.h>
+#include <string.h>
+
+/* single engine */
+struct mutex lock = MUTEX_INITIALIZER;
+
+#define FIRST_PACKET	BIT(30)
+#define NEXT_PACKET	BIT(31)
+
+enum versal_sha3_state {
+	SHA3_STNDBY = 0, SHA3_INIT, SHA3_RUN
+};
+
+struct update_request {
+	uint8_t *data;
+	size_t len;
+	STAILQ_ENTRY(update_request) link;
+};
+
+struct versal_hash_ctx {
+	struct crypto_hash_ctx hash_ctx;
+	enum versal_sha3_state state;
+	STAILQ_HEAD(update_request_list, update_request) req_list;
+};
+
+static const struct crypto_hash_ops versal_ops;
+
+static struct versal_hash_ctx *to_versal_ctx(struct crypto_hash_ctx *ctx)
+{
+	assert(ctx && ctx->ops == &versal_ops);
+
+	return container_of(ctx, struct versal_hash_ctx, hash_ctx);
+}
+
+static TEE_Result do_hash_init(struct crypto_hash_ctx *ctx)
+{
+	struct versal_hash_ctx *c = to_versal_ctx(ctx);
+
+	if (c->state != SHA3_STNDBY)
+		return TEE_ERROR_GENERIC;
+
+	c->state = SHA3_INIT;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result do_hash_update(struct crypto_hash_ctx *ctx,
+				 const uint8_t *data, size_t len)
+{
+	struct versal_hash_ctx *c = to_versal_ctx(ctx);
+	struct update_request *req = NULL;
+
+	/*
+	 * The engine does not have the concept of context, so all update
+	 * requests will be queued and processed at the final stage
+	 */
+	req = malloc(sizeof(*req));
+	if (!req)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	req->data = malloc(len);
+	if (!req->data) {
+		free(req);
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	req->len = len;
+	memcpy(req->data, data, req->len);
+	STAILQ_INSERT_TAIL(&c->req_list, req, link);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result hash_update(struct crypto_hash_ctx *ctx,
+			      const uint8_t *data, size_t len)
+{
+	struct versal_hash_ctx *c = to_versal_ctx(ctx);
+	TEE_Result ret = TEE_SUCCESS;
+	struct versal_mbox_mem p = { };
+	struct versal_cmd_args arg = { };
+	uint32_t init_mask = 0;
+	uint32_t err = 0;
+
+	if (c->state == SHA3_INIT)
+		init_mask = FIRST_PACKET;
+
+	versal_mbox_alloc(len, data, &p);
+
+	arg.ibuf[0].mem = p;
+	arg.data[0] = NEXT_PACKET | init_mask | len;
+	arg.dlen = 1;
+
+	if (versal_crypto_request(VERSAL_SHA3_UPDATE, &arg, &err))
+		ret = TEE_ERROR_GENERIC;
+	else
+		c->state = SHA3_RUN;
+
+	free(p.buf);
```

Comment:
`versal_mbox_free()` helper functions would really be nice. It helps checking balanced malloc/free calls.